### PR TITLE
Implementing Tracing Functionality in Flow Execution

### PR DESCRIPTION
**Pull request by liucheng-ms**
# Description

This PR introduces tracing functionality in the flow execution process. The primary change involves wrapping the original `_exec` function with a new function, `_exec_with_trace`, which starts a new span, enriches it with input and output data, and sets the span status. This enhancement allows us to organize child spans into a flow-level trace, providing a more comprehensive view of flow execution.

Key changes include:

- New Function - `_exec_with_trace`: This function is similar to `_exec`, but includes additional tracing functionality. It starts a new span, enriches it with input and output data, and sets the span status.
- Exception Handling: Exceptions are handled within the context of the new function. In the event of an exception, the span status is set to `StatusCode.ERROR`, and the exception is raised.
- Span Enrichment: The new function enriches the span with input and output data, providing a more detailed view of the flow execution process.
- Status Setting: The span status is set based on the execution of the function. If the function executes successfully, the status is set to `StatusCode.OK`. If an exception occurs, the status is set to `StatusCode.ERROR`.

This enhancement will improve our ability to track and monitor the flow execution process, aiding in debugging and performance optimization efforts.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_core/tracer.py [hunk 1]
```diff
@@ -209,7 +209,7 @@ def enrich_span_with_trace(span, trace):
         span.set_attributes(
             {
                 "framework": "promptflow",
-                "span_type": trace.type,
+                "span_type": f"{trace.type}",
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/tracer.py:212]: How about `trace.type.value`?

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 2]
```diff
@@ -746,6 +743,7 @@ def _get_node_referenced_flow_inputs(
                 node_referenced_flow_inputs[value.value] = flow_inputs[value.value]
         return node_referenced_flow_inputs
 
+    @entry_trace
```
Comment (zzn2) [inline src/promptflow/promptflow/executor/flow_executor.py:746]: Let's do it like `@trace(type=TraceType.FLOW)`, and bind callbacks to the enum to edit the span attributes.

File: src/promptflow/promptflow/_core/tracer.py [hunk 3]
```diff
@@ -382,3 +386,36 @@ async def greetings_async(user_id):
...
+                    "function": func.__name__,
+                    "inputs": serialize_attribute(inputs),
+                }
+            )
+            # invoke function
+            result = func(*args, **kwargs)
+            # extract output from result
+            output = result.output
+            # enrich span with output
+            span.set_attribute("output", serialize_attribute(output))
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/tracer.py:415]: Let's call the `_traced()` method and add a callback (maybe named `span_post_processor` before the span is finished. Then we can implement a callback method for "Flow" type specially.

File: src/promptflow/promptflow/_core/tracer.py (hunk 1 above)
Comment (liucheng-ms) [reply src/promptflow/promptflow/_core/tracer.py:212]: OK

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 4]
```diff
@@ -746,6 +746,70 @@ def _get_node_referenced_flow_inputs(
...
+                span.set_status(StatusCode.OK)
+            except Exception as ex:
+                # set status
+                span.set_status(StatusCode.ERROR, str(ex))
+                # raise exception
+                raise ex
+            else:
+                # return result
+                return result
+
```
Comment (thy09) [inline src/promptflow/promptflow/executor/flow_executor.py:812]: Please add test case

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 5]
```diff
@@ -746,6 +746,70 @@ def _get_node_referenced_flow_inputs(
...
+                Flag to indicate if input validation needed. It is used along with "_raise_ex" to
+                define if exception shall be raised if inputs validation (type check, etc) failed
+                The flag is True for Flow Run, False for bulk run as default
+            allow_generator_output:
+                Flag to indicate if generator output is allowed.
+
+        Returns:
+            LineResult: Line run result
+        """
+        with open_telemetry_tracer.start_as_current_span("promptflow.flow") as span:
```
Comment (zzn2) [inline src/promptflow/promptflow/executor/flow_executor.py:778]: Check whether we can get the flow name

File: src/promptflow/promptflow/executor/flow_executor.py [hunk 6]
```diff
@@ -746,6 +746,70 @@ def _get_node_referenced_flow_inputs(
...
+                )
+                # extract output from result
+                output = result.output
+                # enrich span with output
+                enrich_span_with_output(span, output)
+                # set status
+                span.set_status(StatusCode.OK)
+            except Exception as ex:
+                # set status
+                span.set_status(StatusCode.ERROR, str(ex))
```
Comment (zzn2) [inline src/promptflow/promptflow/executor/flow_executor.py:806]: We don't need this. The otel sdk can do this for us automatically.

File: src/promptflow/promptflow/_core/tracer.py [hunk 7]
```diff
@@ -209,7 +209,7 @@ def enrich_span_with_trace(span, trace):
         span.set_attributes(
             {
                 "framework": "promptflow",
-                "span_type": trace.type,
+                "span_type": trace.type.value,
                 "function": trace.name,
                 "inputs": serialize_attribute(trace.inputs),
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/tracer.py:214]: Shall we remove this line and call `enrich_span_with_input`?

**Review by thy09 — DISMISSED** [review]

File: src/promptflow/promptflow/executor/flow_executor.py (hunk 5 above)
Comment (liucheng-ms) [reply src/promptflow/promptflow/executor/flow_executor.py:778]: Let's leave it as it is.

File: src/promptflow/promptflow/executor/flow_executor.py (hunk 4 above)
Comment (liucheng-ms) [reply src/promptflow/promptflow/executor/flow_executor.py:804]: Will do

**Review by PeiwenGaoMS — DISMISSED** [review]

**Review by zzn2 — APPROVED** [review]

**Review by chjinche — APPROVED** [review]

**Pull request merged**